File: internal/docker/health.go
package docker

import (
	"context"
	"fmt"
	"net/http"
	"time"

	"github.com/hexnickk/jib/internal/config"
)

// HealthResult represents the outcome of a single health check.
type HealthResult struct {
	Endpoint   string
	OK         bool
	StatusCode int
	Error      string
}

// retryIntervals defines the backoff intervals for health check retries.
// 5 attempts: 3s, 6s, 12s, 24s, 48s.
var retryIntervals = []time.Duration{
	3 * time.Second,
	6 * time.Second,
	12 * time.Second,
	24 * time.Second,
	48 * time.Second,
}

// CheckHealth performs HTTP health checks after an initial warmup period.
// For each health check endpoint, it retries with exponential backoff (5 attempts).
// All endpoints must pass for the deploy to be considered healthy.
func CheckHealth(ctx context.Context, checks []config.HealthCheck, warmup time.Duration) []HealthResult {
	// Wait for warmup period
	if warmup > 0 {
		select {
		case <-time.After(warmup):
		case <-ctx.Done():
			results := make([]HealthResult, len(checks))
			for i, check := range checks {
				results[i] = HealthResult{
					Endpoint: fmt.Sprintf("http://localhost:%d%s", check.Port, check.Path),
					OK:       false,
					Error:    ctx.Err().Error(),
				}
			}
			return results
		}
	}

	results := make([]HealthResult, len(checks))
	client := &http.Client{Timeout: 5 * time.Second}

	for i, check := range checks {
		endpoint := fmt.Sprintf("http://localhost:%d%s", check.Port, check.Path)
		result := HealthResult{Endpoint: endpoint}

		for attempt, interval := range retryIntervals {
			if ctx.Err() != nil {
				result.Error = ctx.Err().Error()
				break
			}

			req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
			if err != nil {
				result.Error = err.Error()
				break
			}

			resp, err := client.Do(req)
			if err != nil {
				result.Error = err.Error()
				if attempt < len(retryIntervals)-1 {
					select {
					case <-time.After(interval):
					case <-ctx.Done():
						result.Error = ctx.Err().Error()
					}
					continue
				}
				break
			}
			resp.Body.Close()

			result.StatusCode = resp.StatusCode
			if resp.StatusCode >= 200 && resp.StatusCode < 300 {
				result.OK = true
				result.Error = ""
				break
			}

			result.Error = fmt.Sprintf("unhealthy status: %d", resp.StatusCode)
			if attempt < len(retryIntervals)-1 {
				select {
				case <-time.After(interval):
				case <-ctx.Done():
					result.Error = ctx.Err().Error()
				}
			}
		}

		results[i] = result
	}

	return results
}

// AllHealthy returns true if all health check results are OK.
func AllHealthy(results []HealthResult) bool {
	for _, r := range results {
		if !r.OK {
			return false
		}
	}
	return true
}

File: modules/docker/health.ts
import type { HealthCheck } from '@jib/config'

export interface HealthResult {
  endpoint: string
  ok: boolean
  statusCode?: number
  error?: string
}

/** Retry intervals (ms) matching the Go implementation: 3s, 6s, 12s, 24s, 48s. */
const RETRY_INTERVALS_MS = [3_000, 6_000, 12_000, 24_000, 48_000]

export interface CheckHealthOptions {
  warmupMs?: number
  requestTimeoutMs?: number
  /** Injected for tests. Defaults to global fetch. */
  fetchFn?: typeof fetch
  /** Injected for tests. Defaults to real setTimeout. */
  sleep?: (ms: number) => Promise<void>
  /** Override retry schedule (tests pass `[0, 0, ...]` to stay fast). */
  intervalsMs?: number[]
}

/** Maps one configured health check to the localhost URL jib probes after deploy. */
export function dockerBuildEndpoint(check: HealthCheck): string {
  return `http://localhost:${check.port}${check.path}`
}

/**
 * Runs each health check with exponential backoff retries. Every check must
 * pass for the deploy to be considered healthy.
 */
export async function dockerCheckHealth(
  checks: HealthCheck[],
  opts: CheckHealthOptions = {},
): Promise<HealthResult[]> {
  const fetchFn = opts.fetchFn ?? fetch
  const sleep = opts.sleep ?? ((ms) => new Promise((r) => setTimeout(r, ms)))
  const intervals = opts.intervalsMs ?? RETRY_INTERVALS_MS
  const requestTimeoutMs = opts.requestTimeoutMs ?? 5_000

  if (opts.warmupMs && opts.warmupMs > 0) await sleep(opts.warmupMs)

  const results: HealthResult[] = []
  for (const check of checks) {
    results.push(await probe(check, fetchFn, sleep, intervals, requestTimeoutMs))
  }
  return results
}

async function probe(
  check: HealthCheck,
  fetchFn: typeof fetch,
  sleep: (ms: number) => Promise<void>,
  intervals: number[],
  requestTimeoutMs: number,
): Promise<HealthResult> {
  const endpoint = dockerBuildEndpoint(check)
  let lastError: string | undefined
  let lastStatus: number | undefined
  for (let attempt = 0; attempt < intervals.length; attempt++) {
    try {
      const ctrl = new AbortController()
      const t = setTimeout(() => ctrl.abort(), requestTimeoutMs)
      const res = await fetchFn(endpoint, { signal: ctrl.signal })
      clearTimeout(t)
      lastStatus = res.status
      if (res.status >= 200 && res.status < 300) {
        return { endpoint, ok: true, statusCode: res.status }
      }
      lastError = `unhealthy status: ${res.status}`
    } catch (err) {
      lastError = err instanceof Error ? err.message : String(err)
    }
    if (attempt < intervals.length - 1) await sleep(intervals[attempt] ?? 0)
  }
  const out: HealthResult = { endpoint, ok: false }
  if (lastStatus !== undefined) out.statusCode = lastStatus
  if (lastError !== undefined) out.error = lastError
  return out
}

/** Matches Go: returns true on an empty list (no checks configured = healthy). */
export function dockerAllHealthy(results: HealthResult[]): boolean {
  return results.every((r) => r.ok)
}
